util/cmp/stream.go

```diff
@@ -0,0 +1,255 @@
+			if err == io.EOF {
+				break
+			}
+			return fmt.Errorf("buffer reader error: %w", err)
+		}
+	}
+	return nil
+}
+
+func closeAndDelete(f *os.File) {
+	if f == nil {
+		return
+	}
+	if err := f.Close(); err != nil {
+		log.Warnf("error closing file %q: %s", f.Name(), err)
+	}
+	if err := os.Remove(f.Name()); err != nil {
+		log.Warnf("error removing file %q: %s", f.Name(), err)
+	}
+}
+
+// compressFiles will create a tgz file with all contents of appPath
+// directory excluding the .git folder. Returns the file alongside
+// its sha256 hash to be used as checksum. It is the responsibility
+// of the caller to close the file.
+func compressFiles(appPath string) (*os.File, string, error) {
+	excluded := []string{".git"}
+	appName := filepath.Base(appPath)
+	tempDir, err := files.CreateTempDir()
+	if err != nil {
+		return nil, "", fmt.Errorf("error creating tempDir for compressing files: %s", err)
+	}
+	tgzFile, err := ioutil.TempFile(tempDir, appName)
+	if err != nil {
+		return nil, "", fmt.Errorf("error creating app temp tgz file: %w", err)
+	}
+	hasher := sha256.New()
+	err = files.Tgz(appPath, excluded, tgzFile, hasher)
+	if err != nil {
+		closeAndDelete(tgzFile)
+		return nil, "", fmt.Errorf("error creating app tgz file: %w", err)
+	}
+	checksum := hex.EncodeToString(hasher.Sum(nil))
+	hasher.Reset()
+
+	// reposition the offset to the beginning of the file for proper reads
+	_, err = tgzFile.Seek(0, io.SeekStart)
+	if err != nil {
+		closeAndDelete(tgzFile)
+		return nil, "", fmt.Errorf("error processing tgz file: %w", err)
+	}
+	return tgzFile, checksum, nil
+}
+
+// receiveFile will receive the file from the gRPC stream and save it in the dst folder.
+// Returns error if checksum doesn't match the one provided in the fileMetadata.
+// It is responsibility of the caller to close the returned file.
+func receiveFile(ctx context.Context, receiver StreamReceiver, checksum, dst string) (*os.File, error) {
+	fileBuffer := bytes.Buffer{}
+	hasher := sha256.New()
+	for {
+		if ctx != nil {
+			if err := ctx.Err(); err != nil {
+				return nil, fmt.Errorf("stream context error: %w", err)
+			}
+		}
+		req, err := receiver.Recv()
+		if err != nil {
+			if err == io.EOF {
+				break
+			}
+			return nil, fmt.Errorf("stream Recv error: %w", err)
+		}
+		f := req.GetFile()
+		if f == nil {
+			return nil, fmt.Errorf("stream request file is nil")
+		}
+		_, err = fileBuffer.Write(f.Chunk)
+		if err != nil {
+			return nil, fmt.Errorf("error writing file buffer: %w", err)
+		}
+		_, err = hasher.Write(f.Chunk)
+		if err != nil {
+			return nil, fmt.Errorf("error writing hasher: %w", err)
+		}
+	}
+	if hex.EncodeToString(hasher.Sum(nil)) != checksum {
+		return nil, fmt.Errorf("file checksum validation error")
+	}
+
+	file, err := ioutil.TempFile(dst, "")
+	if err != nil {
+		return nil, fmt.Errorf("error creating file: %w", err)
+	}
+	_, err = fileBuffer.WriteTo(file)
+	if err != nil {
+		return nil, fmt.Errorf("error writing file: %w", err)
```

Comment:
I see a fix for the error below, but looks like this still needs a closeAndDelete?

Reply:
fixed